File: formcraft-plm/formcraft-plm/src/main/java/fr/formcraft/repo/jpa/StockLotRepository.java
package fr.formcraft.repo.jpa;

import fr.formcraft.model.entity.StockLot;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.math.BigDecimal;
import java.util.List;

@Repository
public interface StockLotRepository extends JpaRepository<StockLot, Long> {

    @Query("SELECT sl FROM StockLot sl LEFT JOIN FETCH sl.product WHERE sl.product.id = :productId ORDER BY sl.expiryDate ASC NULLS LAST")
    List<StockLot> findByProductId(@Param("productId") Long productId);

    @Query("SELECT sl FROM StockLot sl LEFT JOIN FETCH sl.product LEFT JOIN FETCH sl.supplier ORDER BY sl.receivedAt DESC")
    List<StockLot> findAllWithDetails();

    @Query("SELECT COALESCE(SUM(sl.quantityOnHand), 0) FROM StockLot sl WHERE sl.product.id = :productId AND sl.status = 'ACTIVE'")
    BigDecimal totalOnHandForProduct(@Param("productId") Long productId);
}

File: app/repositories/stock_lot_repository.py
from sqlalchemy import func, select
from sqlalchemy.orm import Session, joinedload

from app.models.stock_lot import StockLot


def find_by_id(db: Session, lot_id: int) -> StockLot | None:
    return db.get(StockLot, lot_id)


def find_by_product_id(db: Session, product_id: int) -> list[StockLot]:
    return list(
        db.execute(
            select(StockLot)
            .options(joinedload(StockLot.product))
            .where(StockLot.product_id == product_id)
            .order_by(StockLot.expiry_date.asc().nulls_last())
        ).scalars().unique().all()
    )


def find_all_with_details(db: Session, offset: int = 0, limit: int = 20) -> list[StockLot]:
    return list(
        db.execute(
            select(StockLot)
            .options(joinedload(StockLot.product), joinedload(StockLot.supplier))
            .order_by(StockLot.received_at.desc())
            .offset(offset).limit(limit)
        ).scalars().unique().all()
    )


def count_all(db: Session) -> int:
    return db.execute(select(func.count(StockLot.id))).scalar_one()


def total_on_hand_for_product(db: Session, product_id: int) -> float:
    result = db.execute(
        select(func.coalesce(func.sum(StockLot.quantity_on_hand), 0))
        .where(StockLot.product_id == product_id, StockLot.status == "ACTIVE")
    ).scalar_one()
    return float(result)


def save(db: Session, lot: StockLot) -> StockLot:
    db.add(lot)
    db.flush()
    db.refresh(lot)
    return lot
